### PR TITLE
Dev UI 2

### PR DESCRIPTION
**Pull request by phillip-kruger**
This is the initial PR for the new Dev UI.

There is still a lot of outstanding work, thus for now the old dev ui is still available. Once all extensions and features is ported the old code can be deleted. 

We can either merge this as is, and do follow up PRs that migrate the extensions, or add commits to this PR. I'll leave that up to the reviewers to decide.

You can see the new way of taking part in the Dev UI here (in progress):  https://youtube.com/playlist?list=PLLEUmLgWP9osVhex2HeEovY3a2bJUPbSV

You can also include this dummy extension:

https://github.com/phillip-kruger/quarkus-jokes

```
<dependency>
        <groupId>io.quarkiverse.jokes</groupId>
        <artifactId>quarkus-jokes</artifactId>
        <version>1.0.0-SNAPSHOT</version>
</dependency
```
(it might be useful to publish the above somewhere)

Any example application that includes above, for instance: https://github.com/phillip-kruger/devui-example will show the jokes extension:

You can access the new ui under [/q/dev-ui](http://localhost:8080/q/dev-ui) (note the -ui)

example

**Core functionality to be migrated:**

- [x] Extensions page
- [ ] Configuration page
- [ ] Continuous testing page
- [ ] DevServices page
- [ ] Build steps page

**Core Extensions to be migrated:**

- [ ] arc (in progress - Phillip)
- [x] cache - Clement has the code
- [x] container-image - Clement has the code
- [X] datasource - Clement has the code
- [X] flyway - Clement has the code
- [ ] grpc
- [ ] hibernate-orm
- [ ] hibernate-search-orm-elasticsearch
- [ ] infinispan-client
- [ ] kafka-client
- [ ] kafka-streams
- [ ] kubernetes/openshift
- [ ] kubernetes/vanilla
- [ ] liquibase
- [X] micrometer - Clement has the code
- [ ] oidc
- [ ] qute
- [ ] resteasy-reactive/quarkus-resteasy-reactive
- [ ] resteasy-reactive/rest-client-reactive
- [ ] scheduler
- [ ] smallrye-fault-tolerance
- [x] smallrye-graphql
- [x] smallrye-health - Clement has the code
- [ ] smallrye-metrics
- [x] smallrye-openapi
- [ ] smallrye-reactive-messaging-rabbitmq
- [x] smallrye-reactive-messaging - Clement has the code
- [ ] ... more ?

**Quarkiverse extensions to be migrated:**

- [ ] quarkus-operator-sdk
- [ ] quarkus-neo4j
- [ ] quarkus-jpastreamer
- [ ] quarkus-jef
- [ ] quarkus-cxf
- [ ] quarkus-minio
- [ ] quarkus-dashbuilder
- [ ] quarkus-ngrok
- [ ] quarkus-logging-manager
- [ ] ... more ?

**Other things to do:**

- [x] Remove Jackson databind and use  Vertx Json (Jackson core)
- [x] Make the Menu section more accessible for extensions.
- [x] Make the bottom drawer more accessible for extensions.
- [x] Add history to server log
- [x] Investigate if dev-ui can be moved to it's own extension (out of vertx-http)
- [x] Add notification controller
- [x] Dynamic card options (web component cards)
- [ ] Documentation
- [ ] Add tests (using [playwright](https://playwright.dev/))
- [ ] Add importmap support for older browsers
- [ ] Limit external to localhost
- [ ] Separate extenalPage into embedded and external
- [ ] JsonRPC should be able to return Uni
- [ ] Complete Server log implementation. 
- [ ] Remove old dev ui.
- [ ] Add centralsync for mvnpm.org
- [ ] Get notification on a hot reload so we can update screen

More details about the new dev UI: https://github.com/quarkusio/quarkus/wiki/Dev-UI-v2

### CODE REVIEW

**Comment by maxandersen** [comment]
cool @phillip-kruger - would be good to know a few things:

1) did you manage to get to a point where extensions using old dev UI can show up in the new or somehow be linked to "old" page? whats the approach you suggest?
2) can an extension use devui without adding anything bug a qute template and/or the glue devui class? (i.e. no js/webjar required for basic extensions) ?

**Comment by phillip-kruger** [comment]
1) Not yet no. 
2) Yes you can just use Java or Qute for basic pages

**Comment by gsmet** [comment]
> 1. Not yet no.

If it's too complex, maybe we should focus on porting things to the new Dev UI.

**Comment by phillip-kruger** [comment]
> > 1. Not yet no.
> 
> If it's too complex, maybe we should focus on porting things to the new Dev UI.

Yea, I am not sure how complex it would be, I have not yet had time to look into it.

**Comment by maxandersen** [comment]
https://github.com/phillip-kruger/quarkus-jokes is giving me 404 - moved?

**Comment by maxandersen** [comment]
> https://github.com/phillip-kruger/quarkus-jokes is giving me 404 - moved?

was private. now available.

new dev-ui is at `/q/dev-ui`.

currently does not work on safari,firefox nor edge (blocker issue) - only seem to load on chrome.

**Comment by maxandersen** [comment]
going to `/q/dev-ui-configuration` just sits stuck on `Loading Configuration'

**Comment by phillip-kruger** [comment]
> going to `/q/dev-ui/*` just sits stuck on `Loading Configuration|continous testing|dev services|build steps'

Yes, configuration is not implemented yet (see checklist in PR)

**Comment by maxandersen** [comment]
looking at the jokes repo the feature of "external HTML" is cool; but I would not like us to have that available in devui as is as you can literally fake other parts of the devui making it a security issue.

**Comment by phillip-kruger** [comment]
> looking at the jokes repo the feature of "external HTML" is cool; but I would not like us to have that available in devui as is as you can literally fake other parts of the devui making it a security issue.

We can limit that to localhost (so external is always also you - like swagger-ui)

**Comment by maxandersen** [comment]
qute rendering - is it statically compiled qute or something else? 
not sure what current dev ui model does but we should enable it if possible.

**Comment by maxandersen** [comment]
suggestion: separate notion of externalHtml into embedHtml that is allowed only on quarkus dev hosts (incl. future mgmt port) and externalHtml will explicitly show a "url banner" to make it clear it is not part of the main app or even force a separate open.

File: extensions/vertx-http/deployment/src/main/java/io/quarkus/devui/deployment/BuiltTimeContentProcessor.java [hunk 1]
```diff
@@ -0,0 +1,461 @@
...
+
+        internalBuildTimeData.addBuildTimeData("extensions", response);
+
+        // Sections Menu
+        // TODO: Get this from PageBuildItem
+        List<MenuItem> menuItems = List.of(new MenuItem("qwc-extensions", "puzzle-piece", true),
+                new MenuItem("qwc-configuration", "sliders"),
+                new MenuItem("qwc-continuous-testing", "flask-vial"),
+                new MenuItem("qwc-dev-services", "wand-magic-sparkles"),
+                new MenuItem("qwc-build-steps", "hammer"));
```
Comment (phillip-kruger) [inline extensions/vertx-http/deployment/src/main/java/io/quarkus/devui/deployment/BuiltTimeContentProcessor.java:340]: @cescoffier f.y.i

**Comment by gsmet** [comment]
@phillip-kruger let me know when I can rebase his one.

**Comment by phillip-kruger** [comment]
Thanks @gsmet , I still have an issue with one test case, I'll let you know

**Review by gsmet — CHANGES_REQUESTED** [review]
Let's make sure we integrate https://github.com/quarkusio/quarkus/pull/31189 first and rebase on top of it as we need this change to be properly documented and not hidden in an unrelated PR.
If CI is greenish in this PR, I will do the rebase tomorrow morning.

**Comment by phillip-kruger** [comment]
@gsmet I merged @mkouba 's PR and rebased this PR .

**Comment by phillip-kruger** [comment]
@gsmet @cescoffier we have a green PR!! Now just someone who can review :)

**Review by gsmet — APPROVED** [review]
Let's merge and iterate in the `main` branch.

Very nice work @phillip-kruger , thanks for your dedication.

File: extensions/resteasy-classic/resteasy/deployment/src/test/java/io/quarkus/resteasy/test/ClasspathResourceTestCase.java [hunk 2]
```diff
@@ -19,10 +20,26 @@ public class ClasspathResourceTestCase {
 
     @Test
     public void testRootResource() {
-        RestAssured.when().get("/other/hello.txt").then().body(Matchers.is("hello"));
-        RestAssured.when().get("/stuff.html").then().body(Matchers.is("stuff"));
-        RestAssured.when().get("/index.html").then().body(Matchers.is("index"));
-        RestAssured.when().get("/").then().body(Matchers.is("index"));
+
+        RestAssured.get("/other/hello.txt").then()
+                .log().all()
```
Comment (gsmet) [inline extensions/resteasy-classic/resteasy/deployment/src/test/java/io/quarkus/resteasy/test/ClasspathResourceTestCase.java:25]: We whould probably remove the `.log().all()` in there in a follow-up PR.

File: extensions/resteasy-classic/resteasy/deployment/src/test/java/io/quarkus/resteasy/test/files/StaticFileWithResourcesHttpRootTest.java [hunk 3]
```diff
@@ -29,14 +29,17 @@ public class StaticFileWithResourcesHttpRootTest {
     public void test() {
 
         RestAssured.get("/web/index.html").then()
+                .log().all()
```
Comment (gsmet) [inline extensions/resteasy-classic/resteasy/deployment/src/test/java/io/quarkus/resteasy/test/files/StaticFileWithResourcesHttpRootTest.java:32]: Same here.

File: extensions/resteasy-classic/resteasy/deployment/src/test/java/io/quarkus/resteasy/test/ClasspathResourceTestCase.java (hunk 2 above)
Comment (gsmet) [reply extensions/resteasy-classic/resteasy/deployment/src/test/java/io/quarkus/resteasy/test/ClasspathResourceTestCase.java:25]: I will do it, I noticed a few other small things that I will address.

**Pull request merged**